### ISSUE TITLE
#5113 with generic classes

### ISSUE THREAD

**Issue by sylee957** [microsoft/pyright#5208]
**Describe the bug**

Running the pyright on

```python
from collections.abc import Callable
from dataclasses import dataclass
from typing import Generic, TypeVar, TypeVarTuple

X = TypeVar("X")
Y = TypeVar("Y")
Z = TypeVar("Z")
Xs = TypeVarTuple("Xs")
Ys = TypeVarTuple("Ys")


@dataclass(frozen=True)
class Tree(Generic[X, Y]):
    left: X
    right: Y


def star(f: Callable[[X], Y]) -> Callable[[*tuple[X, ...]], tuple[Y, ...]]:
    def wrapped(*xs: X):
        return tuple(map(f, xs))

    return wrapped


def lift(
    f: Callable[[*Xs], tuple[*Ys]]
) -> Callable[[Tree[Z, tuple[*Xs]]], Tree[Z, tuple[*Ys]]]:
    def wrapped(x: Tree[Z, tuple[*Xs]]) -> Tree[Z, tuple[*Ys]]:
        return Tree(x.left, f(*x.right))

    return wrapped


def test(
    f: Callable[[X], Y]
) -> Callable[[Tree[Z, tuple[X, ...]]], Tree[Z, tuple[Y, ...]]]:
    return lift(star(f))

```

Gives the error
```
- error: Argument of type "(*tuple[X@test, ...]) -> tuple[Y@test, ...]" cannot be assigned to parameter "f" of type "(*Xs@lift) -> tuple[*Ys@lift]" in function "lift"
    Type "(*tuple[X@test, ...]) -> tuple[Y@test, ...]" cannot be assigned to type "(*Xs@lift) -> tuple[*Ys@lift]"
      Parameter 1: type "*Xs@lift" cannot be assigned to type "*tuple[X@test, ...]"
        "object*" is incompatible with "*tuple[X@test, ...]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

While using the workaround to assign to temporary variable fixes the problem

```python
def test(
    f: Callable[[X], Y]
) -> Callable[[Tree[Z, tuple[X, ...]]], Tree[Z, tuple[Y, ...]]]:
    return (_ := lift(star(f)))
```

This is similar to the trick used in #5113

**VS Code extension or command-line**

I'm using pyright `1.1.311` with `strict` option.

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
